Adding Python 3  functionality

aabawer opened the pull request:
This PR contains the core patches required for sanlock python 3 compatability.
We still need to add the path converter and handle resource and lockspace names a bytes

nirs commented on tests/python_test.py:
```diff
@@ -7,6 +7,7 @@
 import io
 import struct
 import time
+import six
```
six is 3rd party module, so like to import these in the second group like pytest.

nirs commented on tests/python_test.py:
```diff
@@ -7,6 +7,7 @@
 import time
+import six
 
 import pytest
```
Move six here.

nirs commented on tests/python_test.py:
```diff
@@ -114,14 +132,15 @@ def test_read_lockspace_4k_invalid_sector_size(sanlock_daemon, user_4k_path):
-def test_write_resource(tmpdir, sanlock_daemon, size, offset):
-    path = str(tmpdir.join("resources"))
+def test_write_resource(tmpdir, sanlock_daemon, filename, encoding, size, offset):
+    path = util.generate_path(tmpdir, filename, encoding)
```
Nice easy to review change, thanks!

nirs commented on tests/util.py:
```diff
@@ -146,3 +146,15 @@ def check_rindex_entry(entry, name, offset=None, flags=None):
 
     if flags is not None:
         assert e_flags == flags
+
```
Missing commit message explaining why this utility is needed.

nirs commented on python/sanlock.c:
```diff
@@ -1777,6 +1777,34 @@ module_init(PyObject* m)
     return 0;
```
Missing commit message explaining the current status. I think we cannot compile
the module yet with python 3, since we use symbols that were removed in python 3.

nirs commented on python/sanlock.c:
```diff
@@ -128,7 +128,7 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
         if (offset == NULL) {
```
The commit message should explain that:

- In python 2 python code does not care about the type of integers returned from sanlock methods.
- In python 3 there is only one type (int), identical to python 2 long.

nirs commented on python/sanlock.c:
```diff
@@ -128,7 +128,7 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
             res->disks[i].offset = 0;
         } else {
-            res->disks[i].offset = PyInt_AsLong(offset);
+            res->disks[i].offset = PyLong_AsLong(offset);
```
offset is uint64_t, so PyLong_AsUnsignedLongLong()

aabawer replied:
probably should be PyLong_AsUnsignedLong(offset)  since "unsigned long long" is not strongly typed on gcc systems and depends on actual register size, as can be seen in /usr/include/bits/types.h:
```
#if __WORDSIZE == 64
typedef signed long int __int64_t;
typedef unsigned long int __uint64_t;
#else
__extension__ typedef signed long long int __int64_t;
__extension__ typedef unsigned long long int __uint64_t;
#endif
```
Also discussed in https://stackoverflow.com/questions/32198368/unsigned-long-long-conflict-with-uint64-t

nirs replied:
I'm not sure what do you mean by "not strongly typed".

Python does not give you a way to specify uint64_t, only PY_LONG_LONG. I think this is the correct type if you want to convert from/to [u]int64_t, ensuring that your code will be portable.
See http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf

If you are not sure about that, why not ask on python mailing?

aabawer replied:
havent got to the mailing yet, but is seems that PyLong_AsUnsignedLongLong(offset) fails and returns ((ULL)(-1)) when offset object is not explicitly PyLong in python2. PyLong_AsUnsignedLong(offset) doesn't fail when this happens. 
Seems the py2 impl has fallback in PyLong_AsUnsignedLong(x) when x is not PyLong but a PyInt [1],  which PyLong_AsUnsignedLongLong does not possess[2].
[1] https://github.com/certik/python-2.7/blob/c360290c3c9e55fbd79d6ceacdfc7cd4f393c1eb/Objects/longobject.c#L397
[2] https://github.com/certik/python-2.7/blob/c360290c3c9e55fbd79d6ceacdfc7cd4f393c1eb/Objects/longobject.c#L973

nirs replied:
So we have 2 options:
- add a helpers that accept both PyInt and PyLong, similar to is_py_integer.
- support only PyInt input on python 2, since we never supported PyLong.

Since our goal is to have python 3 version, and making python 2 version better is non-goal, I think the second option is best.

To keep python 2 version consistent, we can return also PyInt from sanlcok on py2, keeping current behavior.

aabawer replied:
me think that we'll need helper in both cases 
as we need to call PyInt_XXXX in py2 and PyLong_XXXX in py3.

nirs replied:
@aabawer, I have alternative patch for parsing intput:
https://github.com/nirs/sanlock/commit/d3b7f20589dcdfc895ab4d44d404d12fd7e978bf

If we parse input with "K", we have no problem to accept long values, so there is no problem to return long values.

What do you think?

nirs commented on python/sanlock.c:
```diff
@@ -538,7 +538,7 @@ py_read_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
 
     /* fill the dictionary information: iotimeout */
-    if ((ls_entry = PyInt_FromLong(io_timeout)) == NULL)
+    if ((ls_entry = PyLong_FromLong(io_timeout)) == NULL)
```
io_timeout is uint32_t

aabawer replied:
no Py{Int,Long}_From{UnsignedInt,Int} invented yet.

nirs commented on python/sanlock.c:
```diff
@@ -1070,7 +1070,7 @@ py_acquire(PyObject *self __unused, PyObject *args, PyObject *keywds)
     if (version != Py_None) {
         res->flags |= SANLK_RES_LVER;
-        res->lver = PyInt_AsUnsignedLongMask(version);
+        res->lver = PyLong_AsUnsignedLongMask(version);
```
lver is uint64_t, so it cannot overflow. PyLong_AsUnsignedLongLong() should be correct.

nirs commented on python/sanlock.c:
```diff
@@ -1186,7 +1186,7 @@ py_request(PyObject *self __unused, PyObject *args, PyObject *keywds)
     } else {
         res->flags |= SANLK_RES_LVER;
-        res->lver = PyInt_AsUnsignedLongMask(version);
+        res->lver = PyLong_AsUnsignedLongMask(version);
```
same, uint64_t

Much easier to review and get right now.

nirs commented on python/sanlock.c:
```diff
@@ -81,6 +81,17 @@ is_py_integer(PyObject *obj)
 #endif
 }
 
+/* Note that we must keep the input object alive */
```
This is not detailed enough, and may lead to incorrect usage. How about:

    Return a NUL-terminated representation of the contents of obj.

    obj must be a string object (py2) or Unicode object (py3), otherwise return NULL
    and raise TypeError.

    The returned pointer refers to the internal buffer of string, not a copy. It must not be
    deallocated, and the object must be kept alive while the pointer is used.

We should also make sure that we actually return NULL and raise TypeError in both py2 and py3.

aabawer replied:
> We should also make sure that we actually return NULL and raise TypeError in both py2 and py3.

This is the documented behavior in case of failure:
py2: https://docs.python.org/2/c-api/string.html#c.PyString_AsString
py3: https://docs.python.org/3/c-api/unicode.html#c.PyUnicode_AsUTF8AndSize (PyUnicode_AsUTF8 doc refers to this)

nirs replied:
Since you got the links, lets include them in the comment.

nirs commented on python/sanlock.c:
```diff
@@ -116,7 +127,7 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
             offset = PyTuple_GetItem(tuple, 1);
 
-            p = PyString_AsString(path);
+            p = pystring_as_cstring(path);
```
If this return NULL, the issue is a TypeError.

But in line 142 we have:

    if (p == NULL) {
        __set_exception(EINVAL, "Invalid resource path");
        goto exit_fail;
    }

I think this is leftover from code older code. The only case when p == NULL is we could not get the bytes
from the object, which means the error was already set and we can just return NULL.

This error already exist before this change, so we can fix it later.

nirs commented on python/sanlock.c:
```diff
@@ -1325,7 +1336,7 @@ py_killpath(PyObject *self __unused, PyObject *args, PyObject *keywds)
+        p = pystring_as_cstring(item);
 
         if (p == NULL) {
             __set_exception(EINVAL, "Killpath argument not a string");
```
Same issue - if p is NULL, we already set TypeError, and the SanlockException raised is hides the real error.

We have 2 instances of this issue, lets fix both of them later.

aabawer replied:
one less, one left.

nirs replied:
We can eliminate this change by squashing my patch rewriting this function. I'm not fan of keeping temporary changes in the tree.

nirs replied:
We can resolve both by squashing later patches, but if you don't find an easy way we can leave this as is.

aabawer replied:
later patches only cover the use-case  in __parse_resource out of the three occurrences shown here. so we can leave it as is at this point.

nirs replied:
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -953,7 +953,7 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
 
         /* fill the dictionary information: path */
-        if ((ls_value = PyString_FromString(lss[i].host_id_disk.path)) == NULL)
+        if ((ls_value = PyBytes_FromString(lss[i].host_id_disk.path)) == NULL)
```
This is a path - does not belong to this patch - probably should move to the patch handling paths.

aabawer replied:
we'd like py3 and py2 to compile and test at this point so we better let it be.
we'll have FSConverter patch later for paths and we'll deal with resource path as well.

nirs replied:
We have a patch for handling strings and bytes, this patch is about int/long handling.

aabawer replied:
AFAIK its in this commit - creating PyBytes object out of cstring.
The other commit in this PR regarding cstring is for getting the chars representation  of a Pystring/Pyunicode object, the other way around.

nirs replied:
This handles converting sanlock path (cstring) to python type (unicode).

Using bytes is wrong for path. If you don't wan to mix with the change handling path input, split a path handling path output.

nirs replied:
We can use Py_BuildValue("s", path) to convert the paths until we have a better way.

On py2:
s (string) [char *]
Convert a null-terminated C string to a Python object. If the C string pointer is NULL, None is used.

On py3:
s (str or None) [const char *]
Convert a null-terminated C string to a Python str object using 'utf-8' encoding. If the C string pointer is NULL, None is used.

Which is symmetric to what we accept as input for path.

This can also be used for converting numeric values:

   Py_BuildValue("K", offset);

py2:
K (int) [unsigned long long]
Convert a C unsigned long long to a Python integer object.

py3:
K (long) [unsigned PY_LONG_LONG]
Convert a C unsigned long long to a Python long integer object. Only available on platforms that support unsigned long long.

nirs replied:
Note also the magic {items} format:

{items} (dictionary) [matching-items]
Convert a sequence of C values to a Python dictionary. Each pair of consecutive C values adds one item to the dictionary, serving as key and value, respectively.

Can simplify a lot the code creating dicts and list of dicts. Here is an example from _ssl moduel:

```
return Py_BuildValue("{sisisi}", "x509", x509, "crl", crl, "x509_ca", ca);
```

nirs commented on tests/python_test.py:
```diff
@@ -281,9 +283,9 @@ def test_read_resource_owners_invalid_align_size(tmpdir, sanlock_daemon):
+    (MIN_RES_SIZE, long(0)),
     # Large offset.
-    (LARGE_FILE_SIZE, LARGE_FILE_SIZE - MIN_RES_SIZE),
+    (LARGE_FILE_SIZE, long(LARGE_FILE_SIZE - MIN_RES_SIZE)),
```
Best keep one of the offset as int, since we test different apis in these tests.

aabawer replied:
following discussion about PyInt/PyLong in prev patch we will abandon this patch.
Py3 is long by default, Py2 will stay with Ints as of no-regression action course.

nirs replied:
I have alternative patch for parsing intput:
https://github.com/nirs/sanlock/commit/d3b7f20589dcdfc895ab4d44d404d12fd7e978bf

If we parse input with "K", we have no problem to accept long values, so there is no problem to return long values.

What do you think?

nirs reviewed (changes requested):
Great work, thanks!

aabawer commented:
pushed PR updates:
had some re-factorings & rebasings using the Py_BuildValue to get rid from some Py2 native APIs with their Py3 helpers and make life nicer at general.

nirs commented on python/sanlock.c:
```diff
@@ -217,55 +217,24 @@ set_value_error(const char* format, PyObject* obj)
+                "timestamp", hss[i].timestamp,
+                "io_timeout", hss[i].io_timeout,
+                "flags", hss[i].flags);
+        if (ls_entry == NULL)
```
Awesome.

Does it really fail on python 3 or the commit message needs an update?

aabawer replied:
yes. i intentionaly re-based the dict refactoring changes in my branch and removed some py3 helpers patches expected to be replaced. so at this point there is some py3 compatible functionality missing for build.

nirs commented on python/sanlock.c:
```diff
@@ -890,52 +890,17 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
 
-    /* prepare the dictionary holding the information */
-    if ((ls_list = PyList_New(0)) == NULL)
-        goto exit_fail;
```
How can we add the list if we don't create it first?

nirs replied:
Removing this change in my tree, all the tests pass both on py2 and py3.

Of course we don't have a test for get_lockspaces - we need to add one before we touch this code.

Here is travis build with this fixed, with 4 python versions:
https://travis-ci.org/nirs/sanlock/builds/534274480

aabawer replied:
probably a glitch from all the rebases...
good point!

nirs commented on python/sanlock.c:
```diff
@@ -508,33 +519,16 @@ py_read_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+    ls_info = Py_BuildValue(
+            read_lockspace_pyvalue_format(),
+            "lockspace", ls.name,
+            "iotimeout", io_timeout);
```
Moving the ifdef here will make the code much more clear:

```
#if PY_MAJOR_VERSION == 2
    const char *fmt = "{sssI}";
#else
    const char *fmt = "{sysI}";
#endif

ls_info = Py_BuildValue(fmt, "lockspace", ls.name,  "iotimeout", io_timeout);
```

aabawer replied:
but ugly. we'll do it when py2 will be officialy dead.

nirs commented on python/sanlock.c:
```diff
@@ -589,31 +599,12 @@ py_read_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
-        goto exit_fail;
-    rv = PyDict_SetItemString(rs_info, "version", rs_entry);
-    Py_DECREF(rs_entry);
-    if (rv != 0)
```
Having the format here will be much nicer.

aabawer replied:
see prev reply.

nirs commented on python/sanlock.c:
```diff
@@ -1792,4 +1820,7 @@ initsanlock(void)
+#endif
+
 /* vim: set expandtab shiftwidth=4 tabstop=4 : */
+
```
Why this empty line?

nirs commented on tests/daemon_test.py:
```diff
@@ -11,6 +11,7 @@
 
```
Seems that you need to rebase on my py3 branch, since Pavel patches are already merged there.

nirs replied:
I merged this to my py3 branch, but then I probably removed it by a bad rebase :-). Will push update py3 version soon.

nirs commented on tests/util.py:
```diff
@@ -148,3 +148,15 @@ def check_rindex_entry(entry, name, offset=None, flags=None):
+    path = os.path.join(str(dirname), filename)
+    if encoding:
+        path = path.encode(encoding)
+    return path
```
Ready, I'll take this to my py3 branch.

nirs replied:
Merged into my py3 branch.

nirs commented on tests/python_test.py:
```diff
@@ -116,14 +134,15 @@ def test_read_lockspace_4k_invalid_sector_size(sanlock_daemon, user_4k_path):
-def test_write_resource(tmpdir, sanlock_daemon, size, offset):
-    path = str(tmpdir.join("resources"))
+def test_write_resource(tmpdir, sanlock_daemon, filename, encoding, size, offset):
+    path = util.generate_path(tmpdir, filename, encoding)
```
Ready, will take to my py3 branch.

nirs commented on src/direct.c:
```diff
@@ -339,7 +339,7 @@ static int do_delta_action(int action,
 
```
Needs rebase, already in py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -71,6 +71,16 @@ __set_exception(int en, char *msg)
 }
 
+static unsigned long long
+pyinteger_as_ullmask(PyObject *obj)
```
I like the short notation, but it doe snot work for long (pyinteger_as_l?), so better keep the long name:

    pyinteger_as_unsigned_long_long_mask

aabawer replied:
ull is commonly used, just like the printf notations. but if you insist.

nirs commented on python/sanlock.c:
```diff
@@ -124,44 +124,26 @@ __parse_resource(PyObject *obj, struct sanlk_resource **res_ret)
     res->num_disks = num_disks;
```
This patch runs fine with py3 in my tree, the issue is probably another patch.

aabawer replied:
i re-based it to an earlier point in my branch, to get rid of some py3 helpers.

nirs commented on python/sanlock.c:
```diff
@@ -273,7 +273,7 @@ py_register(PyObject *self __unused, PyObject *args)
         return NULL;
```
Please describe in more details how this behave in py2 and py3.

nirs reviewed (changes requested):
Nice to see so much complex code deleted, adding more functionallity :+1: 

Lets try to remove some temporary changes, replaced by Py_BuildValue in the last patches.

nirs commented:
General note - we don't have tests for sanlock.get_lockspaces() and sanlock.killpath(). We need to add minimal tests before we modify the code.

get_lockspaces() can be tested by adding a call in the test adding lockspace, once before we add (returns []) and once after we add.

killpath can be tested  in a new tests, registering with sanlock (for the fd), and then calling killpath. This should be fast test and we can use FILENAMES for testing how unicode/bytes arguments are handled.

nirs commented:
@aabawer, I added the missing patches to my py3 branch, and push it. Please rebase.

nirs commented on python/sanlock.c:
```diff
@@ -217,55 +217,24 @@ set_value_error(const char* format, PyObject* obj)
-        if (rv != 0)
+        /* fill the dictionary information */
+        ls_entry = Py_BuildValue(
+                "{sKsKsKsIsI}",
```
Doc/extending/extending.rst mention a more readable syntax:

    "{s:K,s:K,s:K,s:I,s:I}"

See https://github.com/python/cpython/blob/2.7/Doc/extending/extending.rst#building-arbitrary-values

Turns out that whitespace is ignored, and this works:

    "{s:K, s:K, s:K, s:I, s:I}"

py2:
https://github.com/python/cpython/blob/951af2d7f140be7beb9cda2bcdd54f820c905e45/Python/modsupport.c#L489

py3.6:
https://github.com/python/cpython/blob/2b9d7abdbd4b41e2c624858f5bc80da59d8a681d/Python/modsupport.c#L414

There is a risk in using poorly documented syntax, so I think it will be good idea to ask about this in python mailing list.

aabawer replied:
seems legit (see second from last, taken from https://github.com/python/cpython/blob/master/Doc/extending/extending.rst):
   Py_BuildValue("")                        None
   Py_BuildValue("i", 123)                  123
   Py_BuildValue("iii", 123, 456, 789)      (123, 456, 789)
   Py_BuildValue("s", "hello")              'hello'
   Py_BuildValue("y", "hello")              b'hello'
   Py_BuildValue("ss", "hello", "world")    ('hello', 'world')
   Py_BuildValue("s#", "hello", 4)          'hell'
   Py_BuildValue("y#", "hello", 4)          b'hell'
   Py_BuildValue("()")                      ()
   Py_BuildValue("(i)", 123)                (123,)
   Py_BuildValue("(ii)", 123, 456)          (123, 456)
   Py_BuildValue("(i,i)", 123, 456)         (123, 456)
   Py_BuildValue("[i,i]", 123, 456)         [123, 456]
   Py_BuildValue("{s:i,s:i}",
                 "abc", 123, "def", 456)    {'abc': 123, 'def': 456}
   Py_BuildValue("((ii)(ii)) (ii)",
                 1, 2, 3, 4, 5, 6)          (((1, 2), (3, 4)), (5, 6))

aabawer replied:
i think we can agree on "{s:K,s:K,s:K,s:I,s:I}" as its part of the doc for now.

aabawer committed:
tests: Add test calls for sanlock.get_lockspaces()

We expect to get acquired/added lockspaces details on list.

Applies to both Py2 and Py3.

aabawer committed:
python: Treat lockspace name argument as PyBytes

Argument parsing is done with "S" formatter now,
resulting with PyBytes object.

Accessing the cstring of PyBytes object is done
with PyBytes_AsString API.

Implications:

- Py2 : will be indifferent for this change as
	ascii string in bytes are indistinguishable.
- Py3 : will require input lockspace name to be set
	explicitly as bytes and raise TypeError
	otherwise.

aabawer committed:
tests: Set lockspaces names to explicit bytes notation

- Py2: wouldn't mind this change.
- Py3: would raise TypeError for lockspace name in str form.

aabawer committed:
python: Treat resource name argument as PyBytes

Argument parsing is done with "S" formatter now,
resulting with PyBytes object.

Accessing the cstring of PyBytes object is done
with PyBytes_AsString API.

Implications:

- Py2 : will be indifferent for this change as
        ascii string in bytes are indistinguishable.
- Py3 : will require input resource name to be set
        explicitly as bytes and raise TypeError
        otherwise.

aabawer committed:
tests: Set resources names to explicit bytes notation

- Py2: wouldn't mind this change.
- Py3: would raise TypeError for resource name in str form.

nirs commented on python/sanlock.c:
```diff
@@ -1792,4 +1820,6 @@ initsanlock(void)
 
+#endif
+
 /* vim: set expandtab shiftwidth=4 tabstop=4 : */
```
Looks fine now.

nirs replied:
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -1208,7 +1218,7 @@ py_request(PyObject *self __unused, PyObject *args, PyObject *keywds)
+        res->lver = pyinteger_as_unsigned_long_long_mask(version);
         if (res->lver == -1) {
             __set_exception(EINVAL, "Unable to convert the version value");
             goto exit_fail;
```
Looks fine.

nirs replied:
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -217,55 +217,24 @@ set_value_error(const char* format, PyObject* obj)
+        if (ls_entry == NULL)
             goto exit_fail;
 
         if (PyList_Append(ls_list, ls_entry) != 0)
```
Looks fine

nirs replied:
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -864,6 +864,16 @@ py_rem_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+    return "{s:s,s:K,s:s,s:K,s:I}";
+#else
+    return "{s:y,s:K,s:s,s:K,s:I}";
+#endif
```
Looks fine, but

aabawer replied:
i think we came to realize at that point that #ifdefs in middle of existing code are much uglier.
if you find that having the format string outside the Py_BuildValue call is hard to follow, i think we can settle the case by using Py_BuildValue helper per problematic Py3-Py2 format use-case. for example here we could use helper like:
`static PyObject*
getlockspaces_pyvalue(const sanlk_lockspace* lss)
{
/* call the Py_BuildValue per Py Version and return the PyObject */
....
}`

nirs replied:
We discussed avoiding repeated usage of same ifdefs, by creating a wrapper hiding the difference between python 2 and 3.

There is nothing wrong in having one ifdef in the function calling Py_BuildValue. The important thing is to have the format string near the arguments.

If we have several places converting sanlock lockspace to dict, we can add a helper like lockspace_to_dict using Py_BuildValue internally, but if we have only once place, I don't see a need for such helper.

nirs commented on python/sanlock.c:
```diff
@@ -895,47 +905,16 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
-        if (rv != 0)
+        /* fill the dictionary information */
+        ls_entry = Py_BuildValue(
+                getlockspaces_pyvalue_format(),
```
The format should be here. It does not make sense to:
- Move the format string far from the code using it
- Add a function to return the format string, you could use a constant

aabawer replied:
see prev reply.

nirs commented on python/sanlock.c:
```diff
@@ -460,6 +460,17 @@ py_write_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
+    return "{s:s,s:I}";
+#else
+    return "{s:y,s:I}";
+#endif
```
Same as the other format "helper"

aabawer replied:
replied.

nirs commented on python/sanlock.c:
```diff
@@ -508,33 +519,16 @@ py_read_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
-        goto exit_fail;
+    /* fill the information dictionary */
+    ls_info = Py_BuildValue(
+            read_lockspace_pyvalue_format(),
```
The format should be here.

aabawer replied:
replied.

nirs commented on python/sanlock.c:
```diff
@@ -306,7 +306,7 @@ py_get_alignment(PyObject *self __unused, PyObject *args)
+    return Py_BuildValue("i", rv);
 }
 
 /* init_lockspace */
```
Looks fine

nirs replied:
Merged into my py3 branch.

nirs commented on tests/python_test.py:
```diff
@@ -305,13 +305,25 @@ def test_add_rem_lockspace(tmpdir, sanlock_daemon, size, offset):
         "ls_name", 1, path, offset=offset, wait=False)
     assert acquired is True
 
+    lls = sanlock.get_lockspaces()
```
lls is a bit too cryptic for me. Can you think about a name that will be more clear.

aabawer replied:
ls_list?

nirs replied:
ls is a shortcut for lockspace, and used in "ls_name". ls_list is not consistent with this usage. How about:

    lockspaces = anlock.get_lockspaces()

This works nicely with existing code, like:

    owners = sanlock.read_resource_owners("ls_name", "res_name", disks)

nirs commented on python/sanlock.c:
```diff
@@ -1546,22 +1551,22 @@ with -EBUSY error in this case.\n\
     Py_BEGIN_ALLOW_THREADS
-    rv = sanlock_set_event(lockspace, &he, flags);
+    rv = sanlock_set_event(PyBytes_AsString(lockspace), &he, flags);
     Py_END_ALLOW_THREADS
```
We don't have tests for this code, need to add minimal tests.

This patch is rather big, lets defer this work for now. I want to merge the first path, where all the code and tests pass on python 3, and work on better types in python 3 and smarter path handling in the next phase.

aabawer replied:
Okay.

nirs reviewed (changes requested):
I reviewed up to:
python: Treat lockspace name argument as PyBytes

Lets finish all the patches up to this patch, before we continue with making python 3 API better.

nirs commented on tests/python_test.py:
```diff
@@ -116,14 +134,15 @@ def test_read_lockspace_4k_invalid_sector_size(sanlock_daemon, user_4k_path):
+    path = util.generate_path(tmpdir, filename, encoding)
     util.create_file(path, size)
     disks = [(path, offset)]
 
```
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -71,6 +71,16 @@ __set_exception(int en, char *msg)
     }
 }
 
+static unsigned long long
```
Fixed to uint64_t, matching sanlock type.

nirs commented on python/sanlock.c:
```diff
@@ -895,47 +895,20 @@ py_get_lockspaces(PyObject *self __unused, PyObject *args, PyObject *keywds)
+                "path", lss[i].host_id_disk.path,
+                "offset", lss[i].host_id_disk.offset,
+                "flags", lss[i].flags);
+        if (ls_entry == NULL)
```
Much nicer like this, thanks!

nirs replied:
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -582,31 +582,16 @@ py_read_resource(PyObject *self __unused, PyObject *args, PyObject *keywds)
+    if (rs_info  == NULL)
         goto exit_fail;
 
     /* success */
```
Merged into my py3 branch.

nirs commented on python/sanlock.c:
```diff
@@ -508,33 +508,20 @@ py_read_lockspace(PyObject *self __unused, PyObject *args, PyObject *keywds)
-    return NULL;
 }
 
 /* read_resource */
```
Merged into my py3 branch.

nirs reviewed (commented):
I merged everything up to:
https://github.com/nirs/sanlock/pull/12/commits/b15170d81cf1ad660fd2c7d89decdfd87ba81159

Please rebase on my py3 branch.

nirs commented:
@aabawer please rebase on py3 branch and remove the patches up to 
python: Use Py_BuildValue("i", ...) to create py integers
which were merged into my branch.

Lets continue this work in a new more focused PR for handling lockspace and resource names on python 3.

aabawer commented:
Rebased.

On Sun, May 19, 2019 at 9:23 PM Nir Soffer <notifications@github.com> wrote:

> @aabawer <https://github.com/aabawer> please rebase on py3 branch and
> remove the patches up to
> python: Use Py_BuildValue("i", ...) to create py integers
> which were merged into my branch.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/12?email_source=notifications&email_token=AL3BPNMGWYZ4GH4SA4I3GJ3PWGLLNA5CNFSM4HNNULY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVXHW5A#issuecomment-493779828>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNNPQ2SFZ5LSEYD47KDPWGLLNANCNFSM4HNNULYQ>
> .
>

aabawer commented on tests/python_test.py:
```diff
@@ -305,13 +305,25 @@ def test_add_rem_lockspace(tmpdir, sanlock_daemon, size, offset):
+
 
 def test_add_rem_lockspace_async(tmpdir, sanlock_daemon):
     path = str(tmpdir.join("ls_name"))
```
not merged into nirs/py3 for a reason?

nirs replied:
I stopped when all the tests passed. This one looks ready now. I think you posted this version while I was merging, and I missed this change.

nirs reviewed (commented):
Looks good in quick look. I need more time to review.

I would be nice if you make the changes in this order:
- add missing test for all calls using lockspace/resource names
- change the test to use always bytes for py3
- keep both b"name" and "name" in py2, ensuring that they are really the same.
  current vdsm code uses "name", and sanlock must be compatible with it when running on python 2.
  vdsm for python 3 will have to change all calls to use b"name".
- If some parameters fail with py3, mark the failing parameter as xfail
- make the changes the module to require bytes in py3, removing possible xfails.

Now that the all the tests are passing, posting a patch breaking the tests is not an option.

nirs commented:
@aabawer can you close this PR and start to a new PR focused on the bytes changes?

aabawer commented:
Ack.

On Mon, May 20, 2019 at 1:34 AM Nir Soffer <notifications@github.com> wrote:

> *@nirs* commented on this pull request.
>
> Looks good in quick look. I need more time to review.
>
> I would be nice if you make the changes in this order:
>
>    - add missing test for all calls using lockspace/resource names
>    - change the test to use always bytes for py3
>    - keep both b"name" and "name" in py2, ensuring that they are really
>    the same.
>    current vdsm code uses "name", and sanlock must be compatible with it
>    when running on python 2.
>    vdsm for python 3 will have to change all calls to use b"name".
>    - If some parameters fail with py3, mark the failing parameter as xfail
>    - make the changes the module to require bytes in py3, removing
>    possible xfails.
>
> Now that the all the tests are passing, posting a patch breaking the tests
> is not an option.
> ------------------------------
>
> In tests/python_test.py
> <https://github.com/nirs/sanlock/pull/12#discussion_r285399930>:
>
> >      sanlock.rem_lockspace("ls_name", 1, path, offset=offset)
>
>      # Once the lockspace is released, we exepect to get False.
>      acquired = sanlock.inq_lockspace(
>          "ls_name", 1, path, offset=offset, wait=False)
>      assert acquired is False
>
> +    lockspaces = sanlock.get_lockspaces()
> +    assert lockspaces == []
> +
>
>  def test_add_rem_lockspace_async(tmpdir, sanlock_daemon):
>      path = str(tmpdir.join("ls_name"))
>
> I stopped when all the tests passed. This one looks ready now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nirs/sanlock/pull/12?email_source=notifications&email_token=AL3BPNNY3GR347RPVDJLMXDPWHIVLA5CNFSM4HNNULY2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOBZBMPQY#pullrequestreview-239257539>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AL3BPNKQO5XI4MMYXFSTKA3PWHIVLANCNFSM4HNNULYQ>
> .
>

aabawer commented:
we'll have another PR for lockspace & resource name bytes
